vms/components/fee/validator.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fee
+
+import (
+	"fmt"
+
+	"github.com/holiman/uint256"
+)
+
+type ValidatorState struct {
+	Current  Gas
+	Target   Gas
+	Capacity Gas
+	Excess   Gas
+	MinFee   GasPrice
+	K        Gas
+}
+
+func (v ValidatorState) CurrentFeeRate() GasPrice {
+	return v.MinFee.MulExp(v.Excess, v.K)
+}
+
+func (v ValidatorState) CalculateContinuousFee(seconds uint64) uint64 {
+	if v.Current == v.Target {
+		return uint64(v.MinFee.MulExp(v.Excess, v.K)) * seconds
+	}
+
+	uint256.NewInt(uint64(v.Excess))
+
+	var totalFee uint64
+	if v.Current < v.Target {
+		secondsTillExcessIsZero := uint64(v.Excess / (v.Target - v.Current))
```

Comment:
If `v.Excess = 5` and `v.Target - v.Current = 2`, then `secondsTillExcessIsZero = 2`. But that isn't correct, as in `2` seconds `v.Excess` will equal `1`.

Reply:
`v.Excess` corresponds to the excess _after_ a block timestamp has elapsed. If `v.Excess` is currently 5, we have already charged the fee for when it is 5.

That means the excess we have not yet charged will be: `[3, 1, 0, 0, ...]` if `v.Target - v.Current = 2`. This then gives us `secondsTillExcessIsZero = 2` and we can optimize the calculation by simply adding `MinFee * (seconds - 2)`.